### DISCUSSION TITLE
An option to refresh lock file update PRs when they get stale

### DISCUSSION THREAD

**Discussion by dominykas** [renovatebot/renovate#24791] [Suggest an Idea]
**What would you like Renovate to be able to do?**

Background discussion: https://github.com/renovatebot/renovate/discussions/9190

At the moment, renovate will not update lock file update PRs unless the "Rebase" checkbox is ticked. Unless the PRs get merged quickly enough, they can go stale. This could result in a delay before the next PR is opened (by default - up to a week).

Renovate should have an option, which would allow to treat it lock file PRs as stale and refresh them without the "Rebase" checbox when enough time has passed.

**Did you already have any implementation ideas?**

A new option under `lockFileMaintenance`: `staleTimeout` (number of days)?

I'd consider working on a PR for this (would probably need some context, as there's usually multiple ways to implement things).

**Comment by rarkins** [comment]
> At the moment, renovate will not update lock file update PRs unless the "Rebase" checkbox is ticked.

Or if it becomes conflicted - can you confirm?

> This could result in a delay before the next PR is opened (by default - up to a week).

Can you elaborate on what you mean by this? Seems like a piece of information is missing here. By "next PR" do you mean the next non-maintenance PR, or do you mean "next maintenance PR if you decide to close or merge the existing stale automerge PR"?

> Renovate should have an option, which would allow to treat it lock file PRs as stale and refresh them without the "Rebase" checbox when enough time has passed.

I'm assuming you mean.. as long as it's within its schedule window, or you would prefer at any time?

**Comment by dominykas** [comment]
> Or if it becomes conflicted - can you confirm?

You mean if I merge some other PR, which modifies the lock file, it should rebase + update the lock file maintenance PR without ticking the box? (During the next maintenance window, presumably?)

> do you mean "next maintenance PR if you decide to close or merge the existing stale automerge PR"?

This. If you merge/close the PR, the next "lock file maintenance" might only get opened in a few days (but the one you just merged is possibly already out of date).

> I'm assuming you mean.. as long as it's within its schedule window, or you would prefer at any time?

Truth be told, since we're running our own nightly job, I'd personally probably set this value to zero (we have a toggle to explicitly request the maintenance at any time), but I'm not sure what makes sense for other people and the SaaS version.

I can see logic and value in both approaches - "refresh the open PR if `staleTimeout` has passed", and "refresh the open PR during the next maintenance window, but only if `staleTimeout` has passed".

**Comment by rarkins** [comment]
> You mean if I merge some other PR, which modifies the lock file, it should rebase + update the lock file maintenance PR without ticking the box? (During the next maintenance window, presumably?)

You made a statement about "renovate will not update lock file update PRs unless..." which I thought was incomplete, so asked you to clarify if you have experience that they are not rebased when conflicted. And yes the use case is that another PR - could be Renovate's or not - modifies the lock file in a conflicting way. If you don't know, then we can simply loosen the PR description to not be so absolute.

FYI Renovate currently rebases conflicted PRs even if out of schedule, as that seems to be the behavior most people expect. I'm not certain if lock file maintenance ones are rebased or not because of this special rebase handling we have, but I think they are.

**Comment by dominykas** [comment]
Oh, cool, I was not aware of this behavior! Thanks for clarifying.

I'll see if I can test it - otherwise sure, agreed - whatever makes sense.

**Comment by HonkingGoose** [comment]
> I'll see if I can test it - otherwise sure, agreed - whatever makes sense.

@dominykas I'm curious, did you manage to test the behavior?

**Comment by dominykas** [comment]
Sorry, this slipped by. Will try to come up with a test case and check right now.

**Comment by dominykas** [comment]
Test steps:

1. Empty repository
2. `npm i @hapi/hoek@9.0.0 && npm shrinkwrap --dev`
3. renovate config: `lockFileMaintenance: { schedule: 'before 8am on Thursday' }`
4. Run renovate
5. Renovate correctly opens a PR to upgrade to 9.1.1 inside the shrinkwrap
6. `npm i @hapi/hoek@9.1.0 && npm shrinkwrap --dev` (this causes a conflict)
7. Re-run renovate

Result: renovate _does_ rebase the PR after a conflict exists.

This bit does work as described 👍

**Comment by HonkingGoose** [comment]
Do we have what we need to label this `status:ready`? Or do we need to think some more?

**Comment by rarkins** [comment]
I think we still need a definite agreement on how refreshing the stale maintenance PRs would work and be configured if necessary

**Comment by HonkingGoose** [comment]
> I think we still need a definite agreement on how refreshing the stale maintenance PRs would work and be configured if necessary

Ok!

---

In the meantime, could a "workaround" help the original poster?

What if you set the lockFileMaintenance` to a daily schedule" within the Renovate config file?
The lockfile PR cannot get stale if you throw it away and get a new one daily...  😉 🙈 

I'm thinking of something like this (do not copy/paste as it's probably wrong...):

```json
{
  "packageRules": [
    {
      "lockFileMaintenance": { "enabled": true },
      "extends": ["schedule:daily"]
    }
  ]
}
```

**Comment by dominykas** [comment]
> What if you set the lockFileMaintenance` to a daily schedule" within the Renovate config file?

I've an option to set it to `at any time` - that's the issue - it does not throw it away (unless we close it manually or tick the checkbox to rebase, etc).

**Comment by arjan-s** [comment]
Hi, writing in from the future here. :)
I just started using Renovate and am trying to find a solution to exactly this issue. I have several (quiet) repos with lock file maintenance PRs that never get updated after initial creation, unless I tick the 'rebase/retry' checkbox.
I'd very much like an option to automate the updating of these PRs, because currently I need to go into every repo and tick the checkbox, wait for Renovate to run, then merge them. Note that I don't want to use automerge because I want to review the updates before merging.

**Reply by jamietanna** [reply]
I'm in agreement that this can be a bit of an awkward case - I thought I'd previously raised something for this, but apparently not

For example, on https://github.com/rootlyhq/rootly-go/pull/102, we may have a `lockFileMaintenance` PR that doesn't get merged, and as the repo itself is fairly low traffic, there's no automagic rebase.

([This branch is configured with `schedule: at any time`](https://github.com/rootlyhq/rootly-go/blob/9acc218d11b64bb59dfc90d4cab8f2834bf07209/renovate.json#L101-L114), but only gets rebases when requested)

Having the option to ensure that a rebase happens on a schedule would be good - I think what we need to decide what that configuration looks like, first - then we can get an Issue raised

**Reply by arjan-s** [reply]
Somewhere online I found that changing the global config setting `lockFileMaintenance.rebaseWhen` to `always` would do this, but sadly it didn't work. :(
Something like that would be nice though.
Or perhaps something new like `lockFileMaintenance.alwaysRegenerate = true`.
